### DISCUSSION TITLE
Use a package with different versions in the same project

### DISCUSSION THREAD

**Discussion by 7tupel** [thheller/shadow-cljs#1151] [Q&A]
Hi, 

Is it possible to use the same (node) package in one project but in two different versions.

I'm evaluating to build a little side project for a hackathon using Clojurescript. Now the complicated thing is that I would like to use the current version of React 18 in that project but on the same time I'm required to use an external (not open source) library with React UI Components that only work with React 16.

Is it possible to somehow configure the compiler to use React 16 when building the code from that external UI Component library but React 18 for anything else?

The UI Components should be encapsulated enough by themselves so that it should not be a problem to use them from another Component using a newer version of React. Especially there is no shared state between the components, everything the Components need are simple props.

I guess it could be possible to compile and bundle the Components into an esm bundle that includes the compiled code of React 16 and then import the code (probably as an Extern) in the main project that in turn uses react 18. But maybe there is a more convenient or more efficient way to do this with shadow-cljs?

cheers Mo

**Comment by thheller** [comment]
Do I understand that correctly that you want to use two different `react` versions in the **same** build? Or is this for two entirely separate builds that share nothing? I guess I don't really understand the question. I'd say no, this is not possible. Not because of shadow-cljs but because of react itself. You can't partially manage the DOM with one version of react and then have it switch to another?

**Reply by 7tupel** [reply]
Ideally I would like to use two versions in the same build but that I believe will not be possible just as you said yourself.

The other way would be to mount a r16 based app inside a r18 app. In this scenario I could create a separate build of the r16 app and use this bundle from the r18 app. That way I could have a r18 component that mounts the other app inside itself (at least I think this should work).

This would surely be complicated, bad to maintain and performance would not be good, especially because I would have to mount not a single r16 app inside the r18 app but several.

Unfortunately the UI library is not open source and the bundle I have is not that nice to use as a starting point to make it r18 compatible

It's probably to build the UI components I need myself to mimic the original ones from the library... :/

Well, it was worth a try.
